src/main/java/com/shapesecurity/salvation2/Policy.java

```diff
@@ -0,0 +1,1126 @@
+package com.shapesecurity.salvation2;
+
+import com.shapesecurity.salvation2.Directives.FrameAncestorsDirective;
+import com.shapesecurity.salvation2.Directives.HostSourceDirective;
+import com.shapesecurity.salvation2.Directives.PluginTypesDirective;
+import com.shapesecurity.salvation2.Directives.ReportUriDirective;
+import com.shapesecurity.salvation2.Directives.SandboxDirective;
+import com.shapesecurity.salvation2.Directives.SourceExpressionDirective;
+import com.shapesecurity.salvation2.URLs.GUID;
+import com.shapesecurity.salvation2.URLs.URI;
+import com.shapesecurity.salvation2.URLs.URLWithScheme;
+import com.shapesecurity.salvation2.Values.Hash;
+import com.shapesecurity.salvation2.Values.Host;
+import com.shapesecurity.salvation2.Values.MediaType;
+import com.shapesecurity.salvation2.Values.RFC7230Token;
+import com.shapesecurity.salvation2.Values.Scheme;
+
+import javax.annotation.Nonnull;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class Policy {
+	// Things we don't preserve:
+	// - Whitespace
+	// - Empty directives or policies (as in `; ;` or `, ,`)
+	// Things we do preserve:
+	// - Source-expression lists being genuinely empty vs consisting of 'none'
+	// - Case (as in lowercase vs uppercase)
+	// - Order
+	// - Duplicate directives
+	// - Unrecognized directives
+	// - Values in directives which forbid them
+	// - Duplicate values
+	// - Unrecognized values
+
+	private List<NamedDirective> directives = new ArrayList<>();
+
+	private SourceExpressionDirective baseUri = null;
+	private boolean blockAllMixedContent = false;
+	private SourceExpressionDirective formAction = null;
+
+	private FrameAncestorsDirective frameAncestors = null;
```

Comment:
I don't understand why some of these directives are stored as fields instead of in the list of directives. It doesn't make sense with the goal to support duplicate directives. Also, duplicate directives are not an error. It looks like if I have two `frame-ancestors` directives that only the first one will be consulted when asking if a URL is allowed.

Reply:
> I don't understand why some of these directives are stored as fields as well as in the list of directives.

So that you can query them without iterating through the whole list of directives.

> It doesn't make sense with the goal to support duplicate directives. Also, duplicate directives are not an error. It looks like if I have two `frame-ancestors` directives that only the first one will be consulted when asking if a URL is allowed.

Yup, that's the behavior required by the specification. Duplicate directives are supported in the sense that they will serialize properly and have the semantics given by the spec (i.e., later ones are ignored).